### DISCUSSION TITLE
Upgrade method from (like) 2.2.1 to 2.2.2?

### DISCUSSION THREAD

**Discussion by hvisage** [poweradmin/poweradmin#419] [Q&A]
Good day,

 I "just" installed 2.2.1 (against a postgresql 14 DB), noticed a problem with the search needing group by extra columns, then as I start the search on github, I noticed 2.2.2.
 
 So yes, I can extract (git clone ; rsync) over the 2.2.1, but then how do I not get install to drop the configurations I've already done, but still "recreate" the needed/required templates (like inc/record.inc.php)?

**Comment by edmondas** [comment]
I'm not sure I fully understand what you're saying. If I understand you correctly, you made some changes to the code and now you want them to be applied to the latest version. If that is the case then it would not be an easy task because now I am doing a lot of changes (refactorings) and even if you create a patch with git/diff it may not apply properly.

If the changes you made resolve some issues, then please create a patch for version 2.1.7 and I will attempt to apply it with the latest code ```diff -ruN poweradmin-2.1.7 poweradmin-2.1.7-your-version >changes.diff```

**Reply by hvisage** [reply]
Lets start with this first: Where is record.inc.php in the git source code

So I fixed a problem in inc/record.inc.php, but can't find it in the git source, it's only on the (post install) web directory:

```
root@poweradmin:~/poweradmin# git status
HEAD detached at v2.2.2
nothing to commit, working tree clean
root@poweradmin:~/poweradmin# find | grep record.inc.php
root@poweradmin:~/poweradmin# cd /srv/wwww
-bash: cd: /srv/wwww: No such file or directory
root@poweradmin:~/poweradmin# cd /srv/www
root@poweradmin:/srv/www# find | grep record.inc.php
./inc/record.inc.php
```

**Reply by edmondas** [reply]
The functions have been moved to lib/DnsRecord.php

**Reply by hvisage** [reply]
:o

That was a tad "unexpected" for patch number update !

But then, how should I have upgraded from 2.2.1 to 2.2.2 ?

**Reply by edmondas** [reply]
PowerAdmin is not a library and was created before semantic versioning was introduced, thus, I don't follow the semver versioning.

No idea what kind of migration issues you have, but functions were moved to another location, they have the same naming. I have to make several changes in order to implement long-awaited functionality and make the code more maintainable.

Alternatively, you can write to me and provide the archive or patches, and I will do my best to help. Of course if the changes are not too secret :)

**Reply by hvisage** [reply]
Thank you for explaining the moved files!

You can do what/how you want and need to, just, it wasn't expected as I picked up a problem I wanted to report on, but couldn't find the record.inc.php so that is why I asked!

 The problem seems to be fixed in 2.2.2, just needed to understand what happen as I didn't expect the files to be moved.
 
 Anycase, I got it working now with Caddy + php-fpm.

**Reply by edmondas** [reply]
It's a very good point. There are some things I should improve:
- provide better release notes (what changed)
- use semantic versioning and follow it

By the way, this is the first time I've heard of this http server.

**Comment by edmondas** [comment]
Closing this discussion as it's outdated. For documentation on the upgrade path, see the [Upgrade Path](https://docs.poweradmin.org/upgrading/#upgrade-path).

Please note that custom code may no longer be compatible due to significant architectural changes made to the codebase. For continued functionality, I recommend either creating a new feature request or implementing your solution based on the current code in the master branch.

Thank you for your understanding and contributions.
